content/blog/nielx/2024-02-05_rfc_coding_guidelines_formatting_class_member_declarations.md

```diff
@@ -0,0 +1,286 @@
+
+### Review of the Table Class Member Style
+
+Haiku has codified a particular style when it comes to declaring classes or structs. The example below is given in the code guidelines:
+
+```c++
+class Foo : public Bar {
+public:
+								Foo(int32);
+	virtual						~Foo();
+
+	virtual void				SomeFunction(SomeType* argument);
+
+	// indent long argument lists by a tab:
+	virtual const char*			FunctionLotsOfArguments(const char* name,
+									const char* path, const char* user);
+
+private:
+			int32				_PrivateMethod();
+	static  int32				_SomeStaticPrivateMethod();
+
+// Redundant private: to separate the fields from the methods:
+private:
+	volatile int32				fMember;
+			const char*			fPointerMember;
+};
+```
+
+As you can see, the resulting code is a table-like formatted class definition. I am unsure whether the rule is actually explicitly spelled out somewhere, so I am going to quote [pulkomandy's definition](https://github.com/owenca/haiku-format/issues/19#issuecomment-852410195) which I use as a guideline myself:
+
+
+> The basic idea is:
+>
+> indented 1 tab: qualifiers (volatile, virtual, static)
+>
+> indented 3 tabs: type of fields and return type of functions
+>
+> indented 8 tabs: function or field name
+
+And indeed, if you look at any of our [public API headers](https://git.haiku-os.org/haiku/tree/headers/os?h=r1beta4), you will see that this generally works out to give decent table-like class definitions. However, a closer look at the headers will also reveal the rule is a bit more flexible than one may think. As PulkoMandy in the same comment clarified:
+
+> In some cases we move the function and field names to a further tab stop (9 or 10) to keep things aligned even when there are long return types. Sometimes we move them back to the left when return types are short, but function or field names are very long.
+
+Whether that is the rule or not is not entirely clear; the coding guidelines are not explicit about this, they just give the example of the main rule as illustrated by the 1 tab, 3 tabs, 8 tabs guideline. For the purpose of this RFC, that explicit definition is taken as the starting point.
+
+### Why the Current Rule Introduces Inconsistencies
+
+The current rule introduces a table-like format for class members. In the (quick) historical search, I did not find the exact origins of this style. From my perspective, it works particularly well under the following conditions:
+
+- It is used in simple, non-nested classes and structs with (mostly) member functions and member data.
+
+- It works best when the class merely declares its members, but does not define them. This works best when there is a clear separation between a definition of a class with the declaration of its members in the header file, and a definition of the members in the source file.
+
+- It works well with standard Haiku's C++98 style, which does not use namespaces, nested classes or exceptions. There is also very limited use of templated types.
```

Comment:
these things are not really related to C++98

namespaces are used (mainly the BPrivate one, but there is some more in the package kit). I don't remember what our rules are for namespaces currently, but for example in WebKit, declarations in namespaces are not indented, so this isn't a problem.

Namespaces, nested classes, templates and exceptions are all available in C++98.

(My main point here is to remove the C++98 mention, we can discuss the other aspects of this probably in the comment of this article once it is merged).

Reply:
Thank you for the proofread. The point of that particular construct was to clarify that the formatting works well if you stick to a particular subset of the language, but I see how it can be construed as an incorrect point (and I will be hammered for it), as well as that the point is made better in other parts as well, so I will take it out.